components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/CertificateBasedTokenBinder.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    @Override
+    public boolean isValidTokenBinding(OAuth2AccessTokenReqDTO oAuth2AccessTokenReqDTO, String bindingReference) {
+
+        String refreshToken = oAuth2AccessTokenReqDTO.getRefreshToken();
+        try {
+            TokenBinding tokenBinding = getBindingFromRefreshToken(refreshToken, OAuth2Util.isHashEnabled());
+            String cnfValue = generateCnfHashValue(oAuth2AccessTokenReqDTO.getHttpServletRequestWrapper());
+
+            if (tokenBinding != null && CERTIFICATE_BASED_TOKEN_BINDER.equals(tokenBinding.getBindingType()) &&
+                    StringUtils.isNotBlank(cnfValue)) {
+                return StringUtils.equals(cnfValue, tokenBinding.getBindingValue()) &&
+                        StringUtils.equals(bindingReference, tokenBinding.getBindingReference());
+            }
+            return false;
+        } catch (IdentityOAuth2Exception e) {
+            return false;
+        }
+    }
+
+    private String generateCnfHashValue(HttpServletRequest request) {
+
+        Base64URL certThumbprint;
+        X509Certificate certificate = null;
+        String headerName = Optional.ofNullable(IdentityUtil.getProperty(OAuthConstants.MTLS_AUTH_HEADER))
+                .orElse(OAuthConstants.CONFIG_NOT_FOUND);
+
+        String certificateInHeader = request.getHeader(headerName);
+        Object certObject = Optional.ofNullable(request.getAttribute(OAuthConstants.JAVAX_SERVLET_REQUEST_CERTIFICATE))
+                .orElse(null);
+
+        if (StringUtils.isNotBlank(certificateInHeader)) {
+            try {
+                certificate = parseCertificate(certificateInHeader);
+            } catch (CertificateException e) {
+                return null;
+            }
+        } else if (certObject instanceof X509Certificate) {
+            certificate = (X509Certificate) certObject;
+        } else if (certObject instanceof X509Certificate[] && ((X509Certificate[]) certObject).length > 0) {
+            List<X509Certificate> x509Certificates = Arrays.asList((X509Certificate[]) certObject);
+            certificate = x509Certificates.get(0);
+        }
+
+        if (certificate != null) {
+            certThumbprint = X509CertUtils.computeSHA256Thumbprint(certificate);
+            return certThumbprint.toString();
+        } else {
+            return null;
+        }
+    }
+
+    private X509Certificate parseCertificate(String content) throws CertificateException {
+
+        byte[] decodedContent = java.util.Base64.getDecoder().decode(StringUtils.trim(content
+                .replaceAll(OAuthConstants.BEGIN_CERT, StringUtils.EMPTY)
+                .replaceAll(OAuthConstants.END_CERT, StringUtils.EMPTY)
+        ));
+
+        return (X509Certificate) CertificateFactory.getInstance(Constants.X509)
+                .generateCertificate(new ByteArrayInputStream(decodedContent));
+    }
+
+    private TokenBinding getBindingFromRefreshToken(String refreshToken, boolean isTokenHashingEnabled)
```

Comment:
Lets add this to the TokenBindingMgtDAO and call that.

Reply:
Fixed in 1189e9d4d13c525be1351555f1270bf00635f164